voila/tornado/execution_request_handler.py

```diff
@@ -48,91 +49,107 @@ async def on_message(
             message = json.loads(message_str)
             action = message.get("action", None)
             payload = message.get("payload", {})
-            if action == "execute":
-                request_kernel_id = payload.get("kernel_id")
-                if request_kernel_id != self._kernel_id:
-                    await self.write_message(
-                        {
-                            "action": "execution_error",
-                            "payload": {"error": "Kernel ID does not match"},
-                        }
+
+            if action != "execute":
+                await self.write_message(
+                    {
+                        "action": "error",
+                        "payload": {"error": f"Unknown action: {action}"},
+                    }
+                )
+                return
+
+            request_kernel_id = payload.get("kernel_id")
+            if request_kernel_id != self._kernel_id:
+                await self.write_message(
+                    {
+                        "action": "execution_error",
+                        "payload": {"error": "Kernel ID does not match"},
+                    }
+                )
+                return
+
+            km = await ensure_async(self.kernel_manager.get_kernel(self._kernel_id))
+            execution_data = self._execution_data.pop(self._kernel_id, None)
+            if execution_data is None:
+                await self.write_message(
+                    {
+                        "action": "execution_error",
+                        "payload": {"error": "Missing notebook data"},
+                    }
+                )
+                return
+
+            nb = execution_data["nb"]
+            self._executor = executor = VoilaExecutor(
+                nb,
+                km=km,
+                config=execution_data["config"],
+                show_tracebacks=execution_data["show_tracebacks"],
+            )
+            executor.kc = await executor.async_start_new_kernel_client()
+
+            # schedule notebook run and return so we don't block other messages
+            IOLoop.current().spawn_callback(self._run_notebook, nb, executor)
```

Comment:
Can we call `add_callback` directly?   https://github.com/tornadoweb/tornado/blob/d30ef74020924be5f214091de95314ca3ce919af/tornado/ioloop.py#L673